libp2p-websocket/src/desktop.rs

```diff
@@ -0,0 +1,253 @@
+	#[inline]
+	pub fn new(inner: T) -> WsConfig<T> {
+		WsConfig { transport: inner }
+	}
+}
+
+impl<T> Transport for WsConfig<T>
+where
+	T: Transport + 'static, // TODO: this 'static is pretty arbitrary and is necessary because of the websocket library
+	T::RawConn: Send, // TODO: this Send is pretty arbitrary and is necessary because of the websocket library
+{
+	type RawConn = Box<AsyncStream>;
+	type Listener = stream::Map<
+		T::Listener,
+		fn((<T as Transport>::ListenerUpgrade, Multiaddr)) -> (Self::ListenerUpgrade, Multiaddr),
+	>;
+	type ListenerUpgrade = Box<Future<Item = Self::RawConn, Error = IoError>>;
+	type Dial = Box<Future<Item = Self::RawConn, Error = IoError>>;
+
+	fn listen_on(
+		self,
+		original_addr: Multiaddr,
+	) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+		let mut inner_addr = original_addr.clone();
+		match inner_addr.pop() {
+			Some(AddrComponent::WS) => {}
+			_ => return Err((self, original_addr)),
+		};
+
+		let (inner_listen, new_addr) = match self.transport.listen_on(inner_addr) {
+			Ok((listen, inner_new_addr)) => {
+				// Need to suffix `/ws` to the listening address.
+				let new_addr = inner_new_addr
+					.encapsulate("/ws")
+					.expect("the /ws suffix is always valid");
+				(listen, new_addr)
+			}
+			Err((transport, _)) => {
+				return Err((
+					WsConfig {
+						transport: transport,
+					},
+					original_addr,
+				));
+			}
+		};
+
+		let listen = inner_listen.map::<_, fn(_) -> _>(|(stream, client_addr)| {
+			// Need to suffix `/ws` to each client address.
+			let client_addr = client_addr
+				.encapsulate("/ws")
+				.expect("the /ws suffix is always valid");
+
+			// Upgrade the listener to websockets like the websockets library requires us to do.
+			let upgraded = stream.and_then(|stream| {
+				stream
+					.into_ws()
+					.map_err(|e| IoError::new(IoErrorKind::Other, e.3))
+					.and_then(|stream| {
+						// Accept the next incoming connection.
+						stream
+							.accept()
+							.map_err(|err| IoError::new(IoErrorKind::Other, err))
+							.map(|(client, _http_headers)| {
+								// Plug our own API on top of the `websockets` API.
+								let framed_data = client
+                                    .map_err(|err| IoError::new(IoErrorKind::Other, err))
+                                    .sink_map_err(|err| IoError::new(IoErrorKind::Other, err))
+                                    .with(|data| Ok(OwnedMessage::Binary(data)))
+                                    .and_then(|recv| {
+                                        match recv {
+                                            OwnedMessage::Binary(data) => Ok(Some(data)),
+                                            OwnedMessage::Text(data) => Ok(Some(data.into_bytes())),
+                                            OwnedMessage::Close(_) => Ok(None),
+                                            // TODO: pings and pongs ; freaking hard
+                                            _ => unimplemented!()
```

Comment:
Either turn this into a real error with "pings and pongs are unimplemented" as the message or mark this struct as experimental (or something similar), we can't have the server explode if you send a ping.